contracts/distribution/subdomains/pricing/ZNSAsymptoticPricing.sol

```diff
@@ -0,0 +1,208 @@
+        priceConfigs[domainHash].minPrice = priceConfig.minPrice;
+        priceConfigs[domainHash].maxLength = priceConfig.maxLength;
+        priceConfigs[domainHash].feePercentage = priceConfig.feePercentage;
+        setPrecisionMultiplier(domainHash, priceConfig.precisionMultiplier);
+
+        _validateConfig(domainHash);
+
+        emit PriceConfigSet(
+            domainHash,
+            priceConfig.maxPrice,
+            priceConfig.minPrice,
+            priceConfig.maxLength,
+            priceConfig.baseLength,
+            priceConfig.precisionMultiplier,
+            priceConfig.feePercentage
+        );
+    }
+
+    function setMaxPrice(
+        bytes32 domainHash,
+        uint256 maxPrice
+    ) external onlyOwnerOrOperator(domainHash) {
+        priceConfigs[domainHash].maxPrice = maxPrice;
+
+        if (maxPrice != 0) _validateConfig(domainHash);
+
+        emit MaxPriceSet(domainHash, maxPrice);
+    }
+
+    function setMinPrice(
+        bytes32 domainHash,
+        uint256 minPrice
+    ) external onlyOwnerOrOperator(domainHash) {
+        priceConfigs[domainHash].minPrice = minPrice;
+
+        _validateConfig(domainHash);
+
+        emit MinPriceSet(domainHash, minPrice);
+    }
+
+    function setBaseLength(
+        bytes32 domainHash,
+        uint256 length
+    ) external onlyOwnerOrOperator(domainHash) {
+        priceConfigs[domainHash].baseLength = length;
+
+        _validateConfig(domainHash);
+
+        emit BaseLengthSet(domainHash, length);
+    }
+
+    function setMaxLength(
+        bytes32 domainHash,
+        uint256 length
+    ) external onlyOwnerOrOperator(domainHash) {
+        priceConfigs[domainHash].maxLength = length;
+
+        if (length != 0) _validateConfig(domainHash);
+
+        emit MaxLengthSet(domainHash, length);
+    }
+
+    function setPrecisionMultiplier(
+        bytes32 domainHash,
+        uint256 multiplier
+    ) public onlyOwnerOrOperator(domainHash) {
+        require(multiplier != 0, "ZNSAsymptoticPricing: precisionMultiplier cannot be 0");
+        require(multiplier <= 10**18, "ZNSAsymptoticPricing: precisionMultiplier cannot be greater than 10^18");
+        priceConfigs[domainHash].precisionMultiplier = multiplier;
+
+        emit PrecisionMultiplierSet(domainHash, multiplier);
+    }
+
+    function setFeePercentage(
+        bytes32 domainHash,
+        uint256 feePercentage
+    ) external onlyOwnerOrOperator(domainHash) {
+        priceConfigs[domainHash].feePercentage = feePercentage;
+        emit FeePercentageSet(domainHash, feePercentage);
+    }
+
+    function setRegistry(address registry_) public override onlyAdmin {
+        _setRegistry(registry_);
+    }
+
+    function setAccessController(address accessController_)
+    external
+    override
+    onlyAdmin {
+        _setAccessController(accessController_);
+    }
+
+    function getAccessController() external view override returns (address) {
+        return address(accessController);
+    }
```

Comment:
still thinking about it. the problem comes from proxies. I could add these to the AAccessControlled, but it would not let us call if from the interface variable contract in external contract. so I added the virtual functions to the abstract to make sure they are always implemented. any ideas on this?

Reply:
Yeah I see why it's this way, as these are the modular options.

I think the question is just what kinds of things are defined at the base level of the registrar and what's in the modular attachements.

Any time something is repeated a lot though, I think it's just evidence we probably could stick to some base case at the higher level.